plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiClassBuilderInterceptor.kt

```diff
@@ -64,31 +56,39 @@ sealed class AbiClassInfo {
  * be stripped. However, if `f` is not callable directly, we only generate a
  * single inline method `f` which should be kept.
  */
-class JvmAbiClassBuilderInterceptor : ClassGeneratorExtension {
+class JvmAbiClassBuilderInterceptor(
+    private val removePrivateTopLevelClasses: Boolean,
+) : ClassGeneratorExtension {
     val abiClassInfo: MutableMap<String, AbiClassInfo> = mutableMapOf()
 
     override fun generateClass(generator: ClassGenerator, declaration: IrClass?): ClassGenerator =
-        AbiInfoClassGenerator(generator, declaration)
-
-    private inner class AbiInfoClassGenerator(
+        AbiInfoClassGenerator(
+            abiClassInfo,
+            removePrivateTopLevelClasses,
+            generator,
+            declaration
+        )
+
+    private class AbiInfoClassGenerator(
+        private val abiClassInfo: MutableMap<String, AbiClassInfo>,
+        removePrivateTopLevelClasses: Boolean,
         private val delegate: ClassGenerator,
         irClass: IrClass?,
     ) : ClassGenerator by delegate {
         private val isPrivateClass = irClass != null && DescriptorVisibilities.isPrivate(irClass.visibility)
-        lateinit var internalName: String
-        var localOrAnonymousClass = false
-        var publicAbi = false
-        val methodInfos = mutableMapOf<Method, AbiMethodInfo>()
-        val maskedMethods = mutableSetOf<Method>() // Methods which should be stripped even if they are marked as KEEP
+        private val classRemovedFromAbi = shouldBeRemovedFromAbi(irClass, removePrivateTopLevelClasses)
+        private lateinit var internalName: String
+        private var localOrAnonymousClass = false
+        private var keepEverything = false
+        private val methodInfos = mutableMapOf<Method, AbiMethodInfo>()
+        private val maskedMethods = mutableSetOf<Method>() // Methods which should be stripped even if they are marked as KEEP
 
         override fun defineClass(
             version: Int, access: Int, name: String, signature: String?, superName: String, interfaces: Array<out String>
         ) {
             // Always keep annotation classes
```

Comment:
What if I change kdoc on the annotation, it would shift source lines, would that break ABI? 

Sounds like annotation should also go through debug info stripping instead of just keeping it as is all the time, I might be wrong, haven't checked.

Reply:
Checked it, debug info doesn't affect annotation. I think it happens due to the fact that they can't have members at all.